docs/YOSEMITE.md

```diff
@@ -0,0 +1,103 @@
+
+An implementation of `Action` declares the facade to the business logic that corresponds to the domain level concern it represents. 
+
+To illustrate that with an example, let’s think of statistics. WooCommerce presents different types of statistics (related to an order, to the number of visitors to the store, of the top earning product, to name a few).
+
+Each of those types of statistics will be declared by a different method in the implementation of `Action` called `StatsAction`.
+
+All current implementations of the `Action` protocol are enumerations.
+
+`Action`s work in close cooperation with `Store`s. An `Action` declares a number of operations that the `Store` will implement.
+
+![Class diagram: Actions](images/yosemite_actions.png)
+
+## Putting it all together. Executing an action to fetch data.
+To continue with the example of the statistics, let’s say we want to present the statistics related to the top earning products.
+
+The sequence for the UI layer to trigger a request to fetch data would be:
+* Initialise the operation that we want to perform. In this case that would mean getting an `Action` (the return value of a `StatsAction.retrieveTopEarnerStats`,) passing it a completion block.
+* Dispatch that operation to the `StoresManager`
+* The `StoresManager` will fin the currently active `StoresManagerState`, forward that operation to the `Dispatcher` aggregated by it, which in turn will forward it to whatever `Store` registered as capable of handling it (in our example, `StatsStore`)
+
+The full listing:
+```
+    func syncTopPerformers(onCompletion: ((Error?) -> Void)? = nil) {
+        let action = StatsAction.retrieveTopEarnerStats(siteID: siteID,
+                                                        granularity: granularity,
+                                                        latestDateToInclude: Date()) { [weak self] error in
+			// Error handling removed for brevity
+            onCompletion?(error)
+        }
+
+        StoresManager.shared.dispatch(action)
+    }
+```
+
+## Rationale
+Why did we implement it that way?
+
+We wanted the design to be highly cohesive and loosely coupled. We also wanted the design to be able to be capable of supporting adding new domain level concerns, without having to rewrite the pre-existing code.
+
+### High cohesion
+The system, is highly cohesive. Each module has a distinct, clearly defined responsibility. `StatsStore` implements business logic for statistics, and `NotificationsStore` for notifications. There are no mixed responsibilities.
+
+### Low coupling
+Each module does not know anything about the internal implementation of other modules. Modules collaborate without knowing too much about each other.
+
+It could be argued that there might be some implicit coupling, due to the convention that for each `Store` there is a corresponding `Action`, but the fact is that Stores and Actions can be completely independent from each other, and the system would still work.
```

Comment:
This sentence needs broken up. It's grammatically correct but hard to understand the what is happening to Store and Action. Simplify please?

Reply:
I think I am going to completely remove it. It is hard to understand, as you mentioned, and I am not sure it actually adds any value.